sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Prefixing.hs

```diff
@@ -20,100 +21,102 @@ module DA.Cli.Damlc.Command.MultiIde.Prefixing (
-    _ -> pure ()
-  pure $ LSP.FromServerMess LSP.SProgress $ notif & LSP.params . LSP.token .~ prefixedToken
-addProgressTokenPrefixToServerMessage _ _ _ msg = pure msg
-
-lookupHomeAndNot :: (LSP.ProgressToken, FilePath) -> ProgressTokenPrefixes -> Maybe LSP.ProgressToken
-lookupHomeAndNot (t, home) m = Map.lookup (t, Nothing) m <|> Map.lookup (t, Just home) m
-
--- Adds the "client" prefix to all workDone and partial result Progress tokens
-addProgressTokenPrefixToClientMessage :: ProgressTokenPrefixesVar -> LSP.FromClientMessage' a -> IO (LSP.FromClientMessage' a)
-addProgressTokenPrefixToClientMessage tokensVar = \case
+-- Progress tokens can be created on both client and subIDEs. They are then reported on by a subIDE, and cancelled by the client.
+--   We need to avoid collisions between tokens created by different subIDEs.
+-- Server created progress tokens use the SWindowWorkDoneProgressCreate notification. We prefix these tokens uniquely to the subIDE that created them.
+--   Handled in addProgressTokenPrefixToServerMessage
+-- Client created progress tokens are created alongside requests, `getProgressLenses` creates lenses for these tokens. We prefix these with `client`
+--   All prefixed tokens also include their original type, so the transformation can be safely reversed.
+--   Handled in addProgressTokenPrefixToClientMessage
+-- Progress messages
+--   Sent from subIDEs, these are always forwarded to the client. However, depending on what created the Token, the prefix will need changing.
+--   If the subIDE created the token, the progress message token will be unprefixed.
+--     We check the first character of the token. Neither server or client can create tokens starting with non-hex character. Our prefixing starts with `i` or `t`
+--     If there is no prefix, we add the prefix of the sending subIDE. No subIDE should ever give progress reports for tokens created by other subIDEs.
+--   If the client created the token, it will have the client prefix. We detect this and remove it, so the client sees the exact name it created.
+--   Handled in addProgressTokenPrefixToServerMessage
+-- Cancel messages
+--   Sent by the client, forwarded to either all or a specific subIDE.
+--   If the token was created by the client, it will have no prefix. We add the prefix and broadcast to all subIDEs
+--     This is safe because it is impossible for a subIDE to generate a token matching the prefixed client token, so the correct subIDE will delete the token, and the rest will ignore
+--   If the token was created by a subIDE, it will have the subIDE's unique prefix. We strip this from the message, and return it to MultiIde.hs.
+--     The message handling there will lookup the IDE matching this prefix and send the message to it. If no IDE exists, we can safely drop the message, as the IDE has been removed.
+
+-- | Convenience type for prefixes
+data ProgressTokenPrefix
+  = SubIDEPrefix T.Text
+  | ClientPrefix
+  | NoPrefix
```

Comment:
Can we break the `NoPrefix` into its own variant / maybe wrapper for ProgressTokenPrefix, so that we can drop the error case from `progressTokenPrefixToText`?